### DISCUSSION TITLE
Ability or documentation on how to add additional fields to generated pydantic models

### DISCUSSION THREAD

**Discussion by destos** [piccolo-orm/piccolo_api#195] [Ideas]
I've been needing to add calculated properties to fields on output types for different tables. At the moment it looks like there is no way to take property on a table class definition and reflect it in the pydantic type and use it in the responses.

**Comment by dantownsend** [comment]
So the situation is something like this?

```python
class Band(Table):
    name = Varchar()
    popularity = Integer()

    @property
    def is_top_band(self):
        return self.popularity > 1000
```

If you're using `create_pydantic_model`, you can subclass it so the properties are used.

Here's a quick experiment I put together:

```python
from piccolo.table import Table
from piccolo.columns import Varchar, Integer
from piccolo.utils.pydantic import create_pydantic_model
from piccolo.engine.sqlite import SQLiteEngine

DB = SQLiteEngine()

class Band(Table, db=DB):
    name = Varchar()
    popularity = Integer()

    @property
    def is_top_band(self):
        return self.popularity > 1000


BandModel = create_pydantic_model(Band)


class CustomBandModel(BandModel):
    is_top_band: bool

    class Config:
        orm_mode = True


if __name__ == '__main__':
    Band.create_table(if_not_exists=True).run_sync()

    band = Band.objects().get_or_create(Band.name == 'Pythonistas', defaults={Band.popularity: 2000}).run_sync()

    model = CustomBandModel.from_orm(band)
    print(model.json())
    # >>> {"name": "Pythonistas", "popularity": 2000, "is_top_band": true}
```

This might work for your own custom endpoints (if using FastAPI or something). But in terms of `PiccoloCRUD` and `PiccoloAdmin` there's no support for this at the moment. We could do something like this:

```python
PiccoloCRUD(table=Band, properties=[Band.popularity])
```

And then those properties would be returned by GET requests.

Was that what you had in mind?

**Reply by destos** [reply]
Thanks for the detailed reply to the request, that is exactly what I'm thinking would be nice to have for PiccoloCRUD. The problem for the the pydantic model generation may be inferring the type from the property?

**Reply by dantownsend** [reply]
Yeah, it's tricky.

It thought we could look at the type annotations. For example:

```python
class Band(Table):
    name = Varchar()
    popularity = Integer()

    @property
    def is_top_band(self) -> bool:
        return self.popularity > 1000

>>> typing.get_type_hints(Band.is_top_band)
TypeError: <property object at 0x123456> is not a module, class, method, or function.
```

But it raises an exception, because properties can't be inspected for some reason.

So we'd have to have something like this with `create_pydantic_model`:

```python
create_pydantic_model(Band, properties={Band.is_top_band: bool})
```

Not a bad solution though.

**Reply by destos** [reply]
Not sure how stable the API is, but you can get the function for the getter with `Band.is_top_band.fget`. Looks like get_type_hints returns useful information then.

**Reply by dantownsend** [reply]
You're right - this seems to work `typing.get_type_hints(Band.is_top_band.fget)`. I think it's stable enough.
